core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -484,13 +485,42 @@ class TransactionStateManager(brokerId: Int,
       responseCallback(responseError)
     }
 
-    replicaManager.appendRecords(
-      newMetadata.txnTimeoutMs.toLong,
-      TransactionLog.EnforcedRequiredAcks,
-      internalTopicsAllowed = true,
-      isFromClient = false,
-      recordsPerPartition,
-      updateCacheCallback)
+    // note that we need to grab the read lock and double check the metadata and coordinator epoch again
+    // until we have appended to log to avoid the scenario that an emigration and immigration happens after
+    // the check when handling the request
+    inReadLock(stateLock) {
```

Comment:
Another way to do this is to have DelayedTxnMarker.tryComplete() check the expected coordinator epoch and transaction state and error out if they don't match. This way, we just need to replace cancelForKey() calls with checkAndComplete() and get rid of cancelForKey().